contracts/cw3-flex-multisig/src/state.rs

```diff
@@ -45,6 +86,53 @@ impl Proposal {
 
         status
     }
+
+    /// update_status sets the status of the proposal to current_status.
+    /// (designed for handler logic)
+    pub fn update_status(&mut self, block: &BlockInfo) {
+        self.status = self.current_status(block);
+    }
+
+    // returns true iff this proposal is sure to pass (even before expiration if no future
+    // sequence of possible votes can cause it to fail)
+    pub fn is_passed(&self, block: &BlockInfo) -> bool {
+        match self.threshold {
+            Threshold::AbsoluteCount { weight_needed } => self.votes.yes >= weight_needed,
+            Threshold::AbsolutePercentage { percentage_needed } => {
+                self.votes.yes >= votes_needed(self.total_weight, percentage_needed)
+            }
+            Threshold::ThresholdQuora { threshold, quroum } => {
+                // this one is tricky, as we have two compares:
+                if self.expires.is_expired(block) {
+                    // * if we have closed yet, we need quorum% of total votes to have voted (counting abstain)
+                    //   and threshold% of yes votes from those who voted (ignoring abstain)
+                    let total = self.votes.total();
+                    let opinions = total - self.votes.abstain;
+                    total >= votes_needed(self.total_weight, quroum)
+                        && self.votes.yes >= votes_needed(opinions, threshold)
+                } else {
+                    // * if we have not closed yet, we need threshold% of yes votes (from 100% voters - abstain)
+                    //   as we are sure this cannot change with any possible sequence of future votes
+                    self.votes.yes
+                        >= votes_needed(self.total_weight - self.votes.abstain, threshold)
```

Comment:
Again, discarding abstain here (or not discarding it above) seems inconsistent to me.

Reply:
I have to think better about this (and probably write some code), but there doesn't need to be a necessity to differentiate between expired or not expired, for knowing if a proposal passed.

If the snapshotted yes votes are greater than a percentage of the initial total weight, and the number of people who voted is greater than a percentage (quorum) of the initial total weight, proposal is passed.

Maybe the distinction is made because of the abstained voters, but, why discount them from the total initial weight?

Too many questions... I propose we discuss tomorrow about this, and then I complete the review.

Reply:
Point here is you can only close early if you assume all missing votes are "no" and you would still pass.

50% threshold and 40% quorum with 20 total weight:

5 yes and 4 no would pass when the proposal is closed (hits 50% threshold of existing voters), but would not be sufficient to close early (what if 3 more nos came in... we cannot call this until we are sure it cannot fail).

We would need 10 yes (and any number of other votes) to call this early.